### PR TITLE
fix: resolve duplicate tool name collision in FC agent runner

### LINKED ISSUE #32450: bug: FC agent runner silently loses data when LLM calls the same tool multiple times
## Self Checks

- [x] This is only a bug report, not a feature request or other type of issue.
- [x] I have searched for existing issues to avoid creating duplicates.

## Dify version

1.13.0

## Cloud or Self Hosted

Self Hosted (Docker)

## Steps to reproduce

1. Create an Agent app with Function Calling mode
2. Add a tool (e.g., `search`) that the LLM might call multiple times
3. Send a prompt that causes the LLM to call the same tool 2+ times in a single thought (e.g., "Search for 'Python tutorial' and also search for 'JavaScript tutorial'")
4. Check the agent thought logs

## ✅ Expected Behavior

All tool calls should be preserved with their individual inputs and outputs. For example:
- `search("Python tutorial")` → result A
- `search("JavaScript tutorial")` → result B

## ❌ Actual Behavior

Only the **last** tool call's data survives. Previous calls are silently overwritten.

In the example above, only `search("JavaScript tutorial")` → result B is stored. The first search call is completely lost.

## Root Cause

The FC agent runner uses **dict comprehensions keyed by tool name** to store tool data. When the same tool is called multiple times, later entries overwrite earlier ones:

```python
# api/core/agent/fc_agent_runner.py

# Line ~128 (streaming path) and ~155 (blocking path):
tool_call_inputs = json.dumps(
    {tool_call[1]: tool_call[2] for tool_call in tool_calls},  # ← dict key collision!
    ensure_ascii=False,
)

# Line ~287-292 (observation and meta):
tool_invoke_meta={
    tool_response["tool_call_name"]: tool_response["tool_invoke_meta"]
    for tool_response in tool_responses  # ← same collision
},
observation={
    tool_response["tool_call_name"]: tool_response["tool_response"]
    for tool_response in tool_responses  # ← same collision
},
```

**4 collision sites** in total — affecting `tool_input`, `observation`, and `tool_invoke_meta` storage.

## Impact

- **Data loss**: Tool call inputs, outputs, and metadata are silently discarded
- **Misleading logs**: Agent thought logs show only one tool call when multiple were actually executed
- **Affects all FC agent apps** where the LLM decides to call the same tool more than once in a single thought

### Degraded agent reasoning in multi-turn conversations

This bug doesn't just affect log display — it **directly degrades agent quality** in subsequent turns.

When building conversation history for the next LLM call (`base_agent_runner.py` → `organize_agent_history`), the system:
1. Reads `agent_thought.tool_input` from DB → creates `AssistantPromptMessage.ToolCall` list
2. Reads `agent_thought.observation` from DB → creates `ToolPromptMessage` list
3. Sends this as conversation history to the LLM

Since the data is already lost at write time, the reconstructed history is **incomplete**. The LLM only sees the last tool call instead of all calls that were actually executed. This causes:

- **Redundant tool calls**: The LLM thinks a previous search was never performed → calls the same tool again → wastes tokens and increases latency
- **Incorrect reasoning**: The LLM makes decisions based on partial information, missing results from earlier tool calls
- **Potential infinite loops**: In edge cases, the agent may repeatedly re-invoke the "missing" tool call, never converging to a final answer

**Example**: LLM calls `search("Python")` then `search("JavaScript")` in thought 1. In thought 2, history only contains `search("JavaScript")`. The LLM believes it never searched for Python → calls `search("Python")` again → wasted round-trip, and the agent's multi-step reasoning is compromised.

## Proposed Fix

PR #32442 switches the storage format from dict to array:

```python
# Before (loses data):
{tool_call[1]: tool_call[2] for tool_call in tool_calls}

# After (preserves all):
[{"name": tool_call[1], "arguments": tool_call[2]} for tool_call in tool_calls]
```

With backward-compatible property accessors that handle both old and new formats — no database migration required.

### PR DESCRIPTION
**Pull request by trongtrandp**
Closes #32450

## Summary
- Fix data loss when LLM calls the same tool multiple times in a single FC agent thought
- Root cause: dict comprehensions keyed by `tool_call_name` silently overwrite entries for duplicate tool names
- Switch storage format from dict `{"tool_name": data}` to array `[{"name": str, "arguments": dict}]` for `tool_input`, `observation`, and `tool_invoke_meta`

## Problem
When the LLM calls the same tool multiple times (e.g., two `search` calls with different queries), the FC agent runner stores results using dict comprehensions:
```python
# Before (loses data):
tool_call_inputs = json.dumps(
    {tool_call[1]: tool_call[2] for tool_call in tool_calls}
)
```
Only the last call's data survives — all previous calls are silently overwritten.

## Solution
Switch to array-based storage that preserves all tool calls:
```python
# After (preserves all):
tool_call_inputs = json.dumps(
    [{"name": tool_call[1], "arguments": tool_call[2]} for tool_call in tool_calls]
)
```

**4 collision sites fixed** in `fc_agent_runner.py`:
1. Streaming path `tool_call_inputs` (line ~128)
2. Blocking path `tool_call_inputs` (line ~155)
3. `tool_invoke_meta` (line ~287)
4. `observation` (line ~290)

## Backward Compatibility
- **No database migration required** — columns are `LongText`, format change is transparent
- Property accessors in `model.py` detect format via `isinstance(data, list)`:
  - `list` → new array format with ordinal key generation (`search`, `search__2`, `search__3`)
  - `dict` → old format logic preserved exactly
- Old data in DB continues to work without any changes

## Files Changed
- `api/core/agent/fc_agent_runner.py` — Storage format dict→array (4 sites)
- `api/models/model.py` — Shared `_parse_array_with_ordinal_keys()` helper + dual-format property accessors
- `api/core/agent/base_agent_runner.py` — Type hints + `organize_agent_history` dual-format support
- `api/services/agent_service.py` — Ordinal key matching via shared helper for agent logs display

## Test Plan
- [x] 21 core unit tests for duplicate tool name handling
- [x] 42 edge case tests (empty names, special characters, unicode, large payloads)
- [x] 18 security tests (injection, overflow, malformed data)
- [x] 32 data flow integrity tests (write→read round-trip, backward compat)
- [x] All 113 tests passing

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>

### CODE REVIEW

File: api/core/agent/base_agent_runner.py [hunk 1]
```diff
@@ -450,43 +450,98 @@ def organize_agent_history(self, prompt_messages: list[PromptMessage]) -> list[P
...
+                                        resp_content = resp_item["output"]
+                                    else:
+                                        resp_content = resp_item
+                                    tool_resp_content = (
+                                        json.dumps(resp_content)
+                                        if isinstance(resp_content, (dict, list))
+                                        else str(resp_content)
+                                    )
+                                elif observation_payload:
+                                    tool_resp_content = observation_payload
```
[the remark this answers is not in the document]
Comment (trongtrandp) [reply api/core/agent/base_agent_runner.py]: Thanks for catching this! You're right — when `tool_responses_parsed` is a list but the index is out of range, falling back to `observation_payload` would send the entire JSON array as a single tool response to the LLM.

Fixed in 09947c8: now when `tool_responses_parsed` is a list, we only index into it — if `idx >= len(...)`, we return an empty string instead. The `elif observation_payload` fallback is now reserved for non-list formats (old dict or unparseable data).

File: api/core/agent/base_agent_runner.py [hunk 2]
```diff
@@ -450,43 +450,98 @@ def organize_agent_history(self, prompt_messages: list[PromptMessage]) -> list[P
...
+                        if isinstance(tool_inputs_parsed, list):
+                            # New array format - iterate by index
+                            for idx, item in enumerate(tool_inputs_parsed):
+                                tool_call_id = str(uuid.uuid4())
+                                if isinstance(item, dict) and "name" in item:
+                                    tool_name = item["name"]
+                                    tool_args = item.get("arguments", {})
+                                else:
+                                    tool_name = tool_names[idx] if idx < len(tool_names) else ""
+                                    tool_args = item if isinstance(item, dict) else {}
```
[the remark this answers is not in the document]
Comment (trongtrandp) [reply api/core/agent/base_agent_runner.py:477]: Good catch — agreed that an empty string tool name could cause validation errors with providers like OpenAI.

Fixed in 09947c8: changed the fallback from `""` to `f"tool_{idx}"` to be consistent with the naming convention in `model.py`.

File: api/models/model.py [hunk 3]
```diff
@@ -2009,16 +2029,36 @@ def tool_inputs_dict(self) -> dict[str, Any]:
...
+                        if isinstance(item, dict) and "name" in item:
+                            name = item["name"]
+                            args = item.get("arguments", {})
+                        else:
+                            name = tools[i] if i < len(tools) else f"tool_{i}"
+                            args = item if isinstance(item, dict) else {}
+                        name_count[name] = name_count.get(name, 0) + 1
+                        key = name if name_count[name] == 1 else f"{name}__{name_count[name]}"
+                        result[key] = args
+                    return result
```
[the remark this answers is not in the document]
Comment (trongtrandp) [reply api/models/model.py]: Agreed — the ordinal key logic was duplicated 4 times across `tool_meta`, `tool_inputs_dict`, `tool_outputs_dict`, and `agent_service.py`.

Fixed in 09947c8: extracted a shared `_parse_array_with_ordinal_keys()` static method on `MessageAgentThought`. All 3 model properties now delegate to it, and `agent_service.py` also calls the same helper to generate ordinal keys. This ensures consistent key generation from a single source of truth.
